File: android/app/src/main/java/com/kaleys/puzzle/Storage.kt
package com.kaleys.puzzle

import android.content.Context
import android.content.SharedPreferences

/**
 * Lightweight persistent settings + collection storage backed by
 * SharedPreferences. Mirrors the iOS Storage helper.
 */
class Storage(context: Context) {

    private val prefs: SharedPreferences =
        context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    var isMuted: Boolean
        get() = prefs.getBoolean(KEY_MUTED, false)
        set(value) = prefs.edit().putBoolean(KEY_MUTED, value).apply()

    /** The timer is hidden by default for a calmer, pressure-free experience. */
    var showTimer: Boolean
        get() = prefs.getBoolean(KEY_SHOW_TIMER, false)
        set(value) = prefs.edit().putBoolean(KEY_SHOW_TIMER, value).apply()

    var tutorialSeen: Boolean
        get() = prefs.getBoolean(KEY_TUTORIAL_SEEN, false)
        set(value) = prefs.edit().putBoolean(KEY_TUTORIAL_SEEN, value).apply()

    fun collection(): MutableSet<String> {
        // Defensive copy — SharedPreferences returns an immutable/shared set.
        return HashSet(prefs.getStringSet(KEY_COLLECTION, emptySet()) ?: emptySet())
    }

    fun addToCollection(key: String) {
        if (key.isEmpty() || key == AnimalImageGenerator.SURPRISE_KEY) return
        val current = collection()
        current.add(key)
        prefs.edit().putStringSet(KEY_COLLECTION, current).apply()
    }

    fun isCollected(key: String): Boolean = collection().contains(key)

    companion object {
        private const val PREFS = "kaleys_puzzle_prefs"
        private const val KEY_MUTED = "kp_muted"
        private const val KEY_SHOW_TIMER = "kp_show_timer"
        private const val KEY_TUTORIAL_SEEN = "kp_tutorial_seen"
        private const val KEY_COLLECTION = "kp_collection"
    }
}

File: ios/KaleysPuzzle/Storage.swift
import Foundation

/// Lightweight persistence for kid-friendly settings, the first-run tutorial
/// flag, and the solved-animal collection. Backed by UserDefaults.
/// Mirrors `storage.js` in the web reference build and `Storage.kt` on Android.
enum Storage {

    private static let defaults = UserDefaults.standard

    private enum Keys {
        static let muted = "kp_muted"
        static let showTimer = "kp_showTimer"
        static let tutorialSeen = "kp_tutorialSeen"
        static let collection = "kp_collection"
    }

    // MARK: - Settings

    static var isMuted: Bool {
        get { defaults.bool(forKey: Keys.muted) }
        set { defaults.set(newValue, forKey: Keys.muted) }
    }

    /// The timer is hidden by default for a calmer, pressure-free experience.
    static var showTimer: Bool {
        get {
            if defaults.object(forKey: Keys.showTimer) == nil { return false }
            return defaults.bool(forKey: Keys.showTimer)
        }
        set { defaults.set(newValue, forKey: Keys.showTimer) }
    }

    // MARK: - Tutorial

    static var tutorialSeen: Bool {
        get { defaults.bool(forKey: Keys.tutorialSeen) }
        set { defaults.set(newValue, forKey: Keys.tutorialSeen) }
    }

    // MARK: - Collection

    /// Set of animal keys the child has completed at least once.
    static func collection() -> Set<String> {
        let arr = defaults.array(forKey: Keys.collection) as? [String] ?? []
        return Set(arr)
    }

    static func addToCollection(_ animalKey: String) {
        var set = collection()
        set.insert(animalKey)
        defaults.set(Array(set), forKey: Keys.collection)
    }

    static func isCollected(_ animalKey: String) -> Bool {
        return collection().contains(animalKey)
    }
}
